### ISSUE TITLE
New rule: Ensure complementary landmarks are top level

### ISSUE THREAD

**Issue by iamrafan** [dequelabs/axe-core#795]
# landmark-complementary-is-top-level

**Description:** Ensure complementary landmarks are top level landmarks

### Tags 
best-practice

## Selector

```<aside>, [role="complementary"]```

### ( none )

1. Check that none of the complementary landmarks are nested inside any HTML5 or role landmarks

**Comment by WilcoFiers** [comment]
Agree, this makes sense. We'll see about including this into our plans for 3.x

**Comment by dylanb** [comment]
What is the basis for this? What is the impact on accessibility if these are not top level?

**Comment by WilcoFiers** [comment]
Confusing page structure. It's pretty low impact I'd say - it is one of the ARIA best practices.

**Comment by dylanb** [comment]
I'm worried about noise - experimental?

**Comment by WilcoFiers** [comment]
Can you elaborate on "noise"?

**Comment by dylanb** [comment]
Developers having to sift through false positives that have dubious impact on users

**Comment by WilcoFiers** [comment]
I wouldn't say that any of the aria best practices are dubious. They are as close to a predefined best practices list as we have.

**Comment by WilcoFiers** [comment]
@dylanb I want to move forward on this, any objection?

**Comment by nickcolley** [comment]
Hello @WilcoFiers !

> Developers having to sift through false positives that have dubious impact on users

I run into this issue a lot. In our most recent case we're using `<aside>` inside the page to show an 'Recommended story'. Sort of an inline advert for another article.

This aligns well with the HTML definition of how to use `<aside>`.

It'd be helpful for us to have any research or evidence that led to introducing this rule, what impact does it have on real users of assistive technologies?

**Comment by BurntCaramel** [comment]
I agree with @nickcolley — I found this rule surprising. Is the solution that I should try to move all my `<aside>` elements outside my `<main>` (I’m not really sure this is possible). Or should I change my `<aside>` to a `<section>` or `<div>`?

I followed the HTML spec’s suggested usage so I find this seems to not align with that? Should the HTML spec say something else?

**Comment by dylanb** [comment]
@WilcoFiers I think we should post some information here on why this is better for users

**Comment by kryp71c** [comment]
MDN complementary role is referenced as an example in the [rule page resources](https://dequeuniversity.com/rules/axe/3.3/landmark-complementary-is-top-level); there are other examples on MDN that actually show the aside element being used inside the main element, including this MDN article about HTML and accessibility: https://developer.mozilla.org/en-US/docs/Learn/Accessibility/HTML

Many of the examples on MDN include usage of asides inside article or section elements. The axe rule states that the only valid use of the aside element is outside of the main element. So, while it is ok to have articles and sections inside the main element, these articles and sections are not allowed to have asides? 
This seems to contradict the suggested semantic use for the aside element as encapsulating tangentially related content "frequently presented as sidebars or call-out boxes". The very first example of an aside on MDN's aside element page is as a pull-quote next to paragraph elements: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/aside ; it seems inconsistent to present this element as the appropriate element to markup pull-quotes or callouts, but not allow it inside the main content, which is the most likely area to feature such devices.
[whatwg declares](https://html.spec.whatwg.org/multipage/sections.html#the-aside-element):

> The element can be used for typographical effects like pull quotes or sidebars, for advertising, for groups of nav elements, and for other content that is considered separate from the main content of the page.

,but unlike the [main element](https://html.spec.whatwg.org/multipage/grouping-content.html#the-main-element) makes no provisions about hierarchically correct usage.

Wai Aria Authoring Practices do indeed state that the aside has to be a top level element: https://www.w3.org/TR/wai-aria-practices-1.1/#html-sectioning-elements, but in step 2 of "[general principles of landmark design](https://www.w3.org/TR/wai-aria-practices-1.1/#general-principles-of-landmark-design)" confusingly states:

> "banner, main, complementary and contentinfo landmarks should be top level landmarks."

, and in the following bullet:

> "Landmark roles can be nested to identify parent/child relationships of the information being presented."

Restricting the aside element to be exclusively top level effectively relegates its use as a sidebar wrapper only.
